stream_filter: Add an option of stream specific linkifiers.

Abhishek-Balaji opened the pull request:
Solves https://github.com/zulip/zulip/issues/4627

As of now, linkifiers are for the entire realm. We can now add a linkifier and assign it only to a specific stream. RealmFilter.stream_id has been added for the same.
In the frontend, we have a typeahead for entering the name.If left blank, the linkifier is assumed to be for the entire realm. But, we switch to stream_id based processing in the backend. (stream_id uses 0 for entire realm filter).
The stream linkifier has priority over the realm one for the same pattern. For the same pattern, we can have multiple linkifiers (as long as they're in different streams). We can have different format_strings in different streams and for the overall realm.

Feel free to ask me if you don't understand any part. It's a bit convoluted, but I've tried to cover as much as possible (might have missed out something). The GIFs should help.

Creating the linkifiers:
![creating](https://user-images.githubusercontent.com/42106909/77897848-d9ec6d80-7297-11ea-9e50-2778834637b4.gif)

Using them:
![exec](https://user-images.githubusercontent.com/42106909/77897852-db1d9a80-7297-11ea-82dd-abbe542524d5.gif)

Only stream (no realm):
![only_stream](https://user-images.githubusercontent.com/42106909/77897856-dc4ec780-7297-11ea-8128-8b9d3d06f7fe.gif)

NOTE 1: I could not think of a way to split it into multiple commits (but the changes are too big). Any suggestions in that area?

NOTE 2: Should I make a commit for updating the official user documentation too? Maybe after this gets approved and/or merged.

showell commented:
It looks like this commit changes a lot of stuff that you didn't intend to change.  Did you have some kind of issue with rebasing or accidentally saving old files?

Abhishek-Balaji committed:
stream_filter: Add an option of stream specific linkifiers.

As of now, linkifiers are for the entire realm. We can now
add a linkifier and assign it only to a specific stream.
RealmFilter.stream_id has been added for the same.
If left blank, the linkifier is assumed to be for the entire
realm. The stream linkifier has priority over the realm one
for the same pattern. For the same pattern, we can have multiple
linkifier objects (as long as they're in different streams).
We can have different format_strings in different streams and
for the overall realm.
In the frontend, we have a typeahead for entering the stream
name. It'll automatically be converted to stream_id in the backend
and processed that way.
Slightly changed working of RealmFilterPatterns in
bugdown.__init__.py to have a list of format_string and stream_id
instead of a single format_string and stream_id for a pattern.
Also changed the OpenAPI docs to accomodate the change.

aero31aero commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1587,12 +1588,23 @@ class RealmFilterPattern(markdown.inlinepatterns.Pattern):
+        msg = self.markdown.zulip_message
+        current_stream = msg.recipient.type_id
+        filter_list_streams = self.data.keys()
+        if current_stream in filter_list_streams:
```
I don't understand this if-else sequence. Why is it not as simple as:

```
if self.stream and not current_stream == self.stream:
    return None
process()
```

I'll leave further comments in CZO.

timabbott commented on zerver/views/realm_filters.py:
```diff
@@ -18,12 +18,19 @@ def list_filters(request: HttpRequest, user_profile: UserProfile) -> HttpRespons
 def create_filter(request: HttpRequest, user_profile: UserProfile, pattern: str=REQ(),
-                  url_format_string: str=REQ()) -> HttpResponse:
+                  url_format_string: str=REQ(),
+                  stream_name: str=REQ(default='')) -> HttpResponse:
```
We should do this by stream_id, not stream_name.

timabbott commented on zerver/views/realm_filters.py:
```diff
@@ -18,12 +18,19 @@ def list_filters(request: HttpRequest, user_profile: UserProfile) -> HttpRespons
+            stream, _, _ = access_stream_by_name(user_profile, stream_name)
+            stream_id = stream.id
+        else:
+            stream_id = 0
```
`None` is generally the right choice of guard for a non-passed argument, and you can do this in REQ using `default=None`
